workers/agent/agent.go

```diff
@@ -0,0 +1,686 @@
+			return
+		}
+
+		// The scheduler picks only runs up which are in state "NotScheduled".
+		// Since the scheduler from the Gaia master instance set the state already to "scheduled",
+		// we have to reset the state here so that the scheduler will pick it up.
+		pipelineRun.Status = gaia.RunNotScheduled
+
+		// Store finally pipeline run
+		if err = a.store.PipelinePutRun(pipelineRun); err != nil {
+			gaia.Cfg.Logger.Error("failed to store pipeline run in store", "error", err.Error(), "pipelinerun", pipelineRunPB)
+			reschedulePipeline()
+			return
+		}
+	}
+
+	// Check if we received work at all
+	if workCounter == 0 {
+		gaia.Cfg.Logger.Trace("got no work from Gaia primary instance. Will try it again after a while...")
+	}
+}
+
+// streamBinary streams the binary in chunks from the remote instance to the given path.
+func (a *Agent) streamBinary(pipelineRunPB *pb.PipelineRun, pipelinePath string) error {
+	// Create a context with timeout
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	ctx = metadata.AppendToOutgoingContext(ctx, idMDKey, a.self.UniqueId)
+	defer cancel()
+
+	// Initiate streaming
+	stream, err := a.client.StreamBinary(ctx, pipelineRunPB)
+	if err != nil {
+		gaia.Cfg.Logger.Error("failed to stream pipeline binary from remote instance", "error", err.Error(), "pipelinerun", pipelineRunPB)
+		return err
+	}
+
+	// Open output file
+	pipelineFile, err := os.Create(pipelinePath)
+	if err != nil {
+		gaia.Cfg.Logger.Error("failed to create new pipeline binary file during remote binary stream", "error", err.Error(), "pipelinerun", pipelineRunPB)
+		return err
+	}
+	defer pipelineFile.Close()
+
+	// Read whole stream
+	for {
+		streamChunk, err := stream.Recv()
+
+		// Check if stream was closed remotely
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			gaia.Cfg.Logger.Error("failed to stream full pipeline binary from remote instance", "error", err.Error(), "pipelinerun", pipelineRunPB)
+			return err
+		}
+
+		// Write chunk to file
+		if _, err := pipelineFile.Write(streamChunk.Chunk); err != nil {
+			gaia.Cfg.Logger.Error("failed to write chunk to local disk during stream binary", "error", err.Error(), "pipelinerun", pipelineRunPB)
+			return err
+		}
+	}
+
+	// Set pipeline executable rights
+	return os.Chmod(pipelinePath, 0766)
+}
+
+// updateWork is function that is periodically called and it is used to
+// send new information about a pipeline run to the Gaia master instance.
+func (a *Agent) updateWork() {
+	// Read all pipeline runs from the store. The number of pipeline runs
+	// should be relatively low since we delete pipeline runs after successful
+	// execution.
+	runs, err := a.store.PipelineGetAllRuns()
+	if err != nil {
+		gaia.Cfg.Logger.Error("failed to load pipeline runs from store", "error", err.Error())
+		return
+	}
+
+	// Send all pipeline runs to the remote primary instance
+	for _, run := range runs {
+		// Transform to protobuf struct
+		runPB := &pb.PipelineRun{
```

Comment:
So, bare with me on this one. What I started to do is instead of creating a struct and assigning the address of it to a variable, I started to do the address of it at the point I want to share it. To clarify, the above code would be like this:

```go
			j := pb.Job{
				UniqueId:    job.ID,
				Title:       job.Title,
				Status:      string(job.Status),
				Description: job.Description,
			}
			runPB.Jobs = append(runPB.Jobs, &j)
```

It's a very subtle change, but makes it clear that at THAT point you are sharing this struct. What say you?

Reply:
I'm wondering if there is some kind of convention? Happy to stick to one solution

Reply:
Not that I know of. I think it's just usage habit. Whatever suits you better I guess. :)